sklearn/neighbors/_nearest_centroid.py

```diff
@@ -255,7 +307,95 @@ def predict(self, X):
         """
         check_is_fitted(self)
 
-        X = self._validate_data(X, accept_sparse="csr", reset=False)
-        return self.classes_[
-            pairwise_distances_argmin(X, self.centroids_, metric=self.metric)
-        ]
+        d = self._decision_function(X)
+        y_pred = self.classes_.take(d.argmax(1))
+        return y_pred
+
+    def _decision_function(self, X):
+        check_is_fitted(self, "centroids_")
+
+        X = self._validate_data(X, reset=False, accept_sparse="csr")
+
+        # check if this works with sparse matrix formats later
+        discriminant_score = np.empty(
+            (X.shape[0], self.classes_.size), dtype=np.float64
+        )
+
+        for cur_class in range(self.classes_.size):
+            Xdist = X - self.centroids_[cur_class, :]
+            Xdist_norm = np.square(Xdist / self.within_class_std_)
```

Comment:
I have to go back to the paper because there is something that I don't get:

- It seems that we don't compute the same quantity compare to the previous implementation
- I am not sure that the previous implementation was following the paper

Reply:
In the previous implementation (current stable version), there was no support for priors. The original authors assumed uniform class priors. Thus calculations can be simplified considerably by skipping the division by the within-class standard deviation (s<sub>j</sub>) and adding/subtracting log class priors (2logπ<sub>k</sub>) as seen here (from the paper):

![image](https://github.com/scikit-learn/scikit-learn/assets/6528822/c761fba4-c5ce-402b-8e64-634bf3e6ab50)

Since I'm adding support for class priors, I added back the full definition.

In addition, the paper is using the argmin definition whereas the textbook ESL is using the argmax definition but is algebraically equivalent.

Reply:
I think we should still use the `pairwise_distances_argmin` for the uniform prior since it will better optimized.

Then, we need to always have the function to take the argmin and not the argmax.
For non-uniform prior, I don't think that we can simplify and we need to get throught the decision function